File: Backend/database/app/utils/oracle_health.py
"""
Shared Oracle health-score formula — backend mirror of
frontend/src/utils/oracleHealth.js, so any future server-side consumer
(e.g. a report/email generator) computes the identical score the dashboard
shows, rather than a third divergent implementation.
"""


def compute_instance_health_score(
    session_pct=None, buf_hit_pct=None, max_ts_pct=None, wait_count=None,
    host_cpu_pct=None, rman_failed=None, archive_gap_count=None, invalid_objects_total=None,
) -> int:
    score = 100
    if (session_pct or 0) > 90:
        score -= 25
    elif (session_pct or 0) > 75:
        score -= 12
    if buf_hit_pct is not None and 0 < buf_hit_pct < 80:
        score -= 25
    elif buf_hit_pct is not None and 0 < buf_hit_pct < 90:
        score -= 12
    if (max_ts_pct or 0) > 95:
        score -= 25
    elif (max_ts_pct or 0) > 85:
        score -= 12
    if (host_cpu_pct or 0) > 90:
        score -= 15
    elif (host_cpu_pct or 0) > 75:
        score -= 8
    if (wait_count or 0) > 20:
        score -= 8
    if (rman_failed or 0) > 0:
        score -= 15
    if (archive_gap_count or 0) > 10:
        score -= 20
    elif (archive_gap_count or 0) > 5:
        score -= 10
    if (invalid_objects_total or 0) > 50:
        score -= 10
    return max(0, score)


def cluster_health_from_nodes(nodes: list) -> str:
    """Topology-aware RAC rollup (§23) — mirrors
    oracle_monitoring_service.oracle_rac_nodes()'s own computation and
    frontend/src/utils/oracleHealth.js's clusterHealthFromNodes()."""
    if not nodes:
        return "not_applicable"
    open_count = sum(1 for n in nodes if str(n.get("instance_status") or "").upper() == "OPEN")
    if open_count == len(nodes):
        return "healthy"
    if open_count == 0:
        return "critical"
    return "warning"

File: frontend/src/utils/oracleHealth.js
/**
 * Shared Oracle health-score formula — used by OracleDashboard and
 * OracleReportsPage so the two never diverge again. Previously each computed
 * its own inline score with different weights and different inputs (e.g. the
 * same instance could score differently on the dashboard vs the report).
 *
 * Every input is OPTIONAL — a signal a caller doesn't have (undefined/null)
 * contributes no penalty, rather than being treated as a bad 0. Only pass a
 * field when it was genuinely measured.
 */
export function computeInstanceHealthScore({
  sessionPct, bufHitPct, maxTsPct, waitCount, hostCpuPct,
  rmanFailed, archiveGapCount, invalidObjectsTotal,
} = {}) {
  let score = 100;
  if (sessionPct > 90) score -= 25;
  else if (sessionPct > 75) score -= 12;
  if (bufHitPct != null && bufHitPct > 0 && bufHitPct < 80) score -= 25;
  else if (bufHitPct != null && bufHitPct > 0 && bufHitPct < 90) score -= 12;
  if (maxTsPct > 95) score -= 25;
  else if (maxTsPct > 85) score -= 12;
  if (hostCpuPct > 90) score -= 15;
  else if (hostCpuPct > 75) score -= 8;
  if (waitCount > 20) score -= 8;
  if (rmanFailed > 0) score -= 15;
  if (archiveGapCount > 10) score -= 20;
  else if (archiveGapCount > 5) score -= 10;
  if (invalidObjectsTotal > 50) score -= 10;
  return Math.max(0, score);
}

/** Topology-aware RAC cluster rollup (§23) — NOT a flat average of node
 * scores: all instances OPEN → healthy, some down → warning, all down →
 * critical. Mirrors the backend's oracle_rac_nodes() computation exactly so
 * the badge always agrees with what the API already decided. */
export function clusterHealthFromNodes(nodes = []) {
  if (!nodes.length) return 'not_applicable';
  const openCount = nodes.filter((n) => String(n.instance_status || '').toUpperCase() === 'OPEN').length;
  if (openCount === nodes.length) return 'healthy';
  if (openCount === 0) return 'critical';
  return 'warning';
}

const BANDS = {
  healthy: { label: 'Healthy', tone: 'success' },
  warning: { label: 'Warning', tone: 'warning' },
  critical: { label: 'Critical', tone: 'danger' },
  not_configured: { label: 'Not Configured', tone: 'neutral' },
  not_applicable: { label: 'N/A', tone: 'neutral' },
};

export function healthBand(status) {
  return BANDS[status] || BANDS.not_applicable;
}

export function bandForScore(score) {
  if (score >= 80) return BANDS.healthy;
  if (score >= 60) return BANDS.warning;
  return BANDS.critical;
}
